packages/neon-api/src/transaction/validator.ts

```diff
@@ -0,0 +1,303 @@
+  SystemFee = 1 << 1,
+  NetworkFee = 1 << 2,
+  Script = 1 << 3,
+  All = ValidUntilBlock | SystemFee | NetworkFee | Script
+}
+
+export type ValidationSuggestion<T> =
+  | {
+      /**
+       * Whether this is auto-fixed by validator
+       * Validator will try to fix the transaction when param `autoFix` is set to TRUE
+       */
+      fixed: boolean;
+      prev: T;
+      suggestion: T;
+    }
+  | {
+      message: T;
+    };
+
+export interface ValidationResult {
+  /**
+   * Whether the transaction is valid after validation
+   */
+  valid: boolean;
+  suggestions?: {
+    validUntilBlock?: ValidationSuggestion<number>;
+    script?: ValidationSuggestion<string>;
+    systemFee?: ValidationSuggestion<u.Fixed8>;
+    networkFee?: ValidationSuggestion<u.Fixed8>;
+  };
+}
+
+/**
+ * A class with functions to validate transaction
+ */
+export class TransactionValidator {
+  /**
+   * Transaction will be validated on this rpc node
+   */
+  public rpcClient: rpc.RPCClient;
+
+  /**
+   * Transaction that will be validated
+   */
+  public transaction: tx.Transaction;
+
+  constructor(rpc: rpc.RPCClient, transaction: tx.Transaction) {
+    this.rpcClient = rpc;
+    this.transaction = transaction;
+  }
+
+  /**
+   * validate transaction attribute - validUntilBlock
+   * @param autoFix will automatically fix transaction if specified as true
+   */
+  async validateValidUntilBlock(autoFix = false): Promise<ValidationResult> {
+    const { validUntilBlock } = this.transaction;
+    const height = await this.rpcClient.getBlockCount();
+    if (
+      validUntilBlock <= height ||
+      validUntilBlock >= height + tx.Transaction.MAX_TRANSACTION_LIFESPAN
+    ) {
+      const suggestion = tx.Transaction.MAX_TRANSACTION_LIFESPAN + height - 1;
+      if (autoFix) {
+        this.transaction.validUntilBlock = suggestion;
+        return {
+          valid: true,
+          suggestions: {
+            validUntilBlock: {
+              fixed: true,
+              prev: validUntilBlock,
+              suggestion
+            }
+          }
+        };
+      }
+      return {
+        valid: false,
+        suggestions: {
+          validUntilBlock: {
+            fixed: false,
+            prev: validUntilBlock,
+            suggestion
+          }
+        }
+      };
+    }
+    return {
+      valid: true
+    };
+  }
+
+  private async _validateIntent(
+    intent: sc.ScriptIntent
+  ): Promise<string | null> {
+    const { scriptHash, operation } = intent;
+    const manifest = await this.rpcClient.getContractState(scriptHash);
+    if (manifest === null) {
+      return `Unknow contract ${scriptHash}`;
+    }
+
+    if (operation) {
+      if (
+        manifest.abi.methods.map(method => method.name).indexOf(operation) < 0
+      ) {
+        return `Unknow method ${operation} for constract ${scriptHash}`;
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * Validate intents
+   */
+  async validateScript(): Promise<ValidationResult> {
```

Comment:
Indeed.
Something that need to be considered: 
1. If we add online testing for script in validateScript and user validate both systemFee and script, the same request would be sent twice.
2. At the begging of testing on neo3 preview testnet, a script will return 'FAULT' when `invokeScript` but actually works in a transaction. I'll try to reproduce this problem and ask someone else if it is a bug.

Reply:
in my experience in neo2, fault can happen because the contract method requires a signature check which an invokescript is unable to provide. However, the script is valid.

Reply:
So shall we use `invokeScript` to validate script? Considering it may return `FAULT` in a valid script.

Reply:
If it fails, we probably set a warning stating that it produced a fault. Autofix should also fail since we cannot accurately estimate the cost if it fails. We could just give the GAS used up to the failure as a baseline in the message.